c/parallel/src/jit_templates/templates/operation.h

```diff
@@ -13,20 +13,60 @@
 #ifndef _CCCL_C_PARALLEL_JIT_TEMPLATES_PREPROCESS
 #  include <cuda/std/cstddef>
 #  include <cuda/std/type_traits>
+#  include <cuda/std/utility>
 
 #  include <cccl/c/types.h>
 #endif
 
 #include "../mappings/operation.h"
 #include "../mappings/type_info.h"
 
+// Helper to define the C function pointer type: void (*)(void*..., void*)
+template <typename Seq>
+struct MakeVoidPtrFuncPtr;
+template <cuda::std::size_t... Is>
+struct MakeVoidPtrFuncPtr<cuda::std::index_sequence<Is...>>
+{
+  // Helper type that is always void*
+  template <cuda::std::size_t>
+  using void_ptr_t = const void*;
+  // Define the function pointer type
+  using type = void (*)(void_ptr_t<Is>..., void*); // Last void* is for result
+};
+
+// Helper to define the C function pointer type for stateful ops: void (*)(void* state, void*..., void*)
+template <typename Seq>
+struct MakeVoidPtrStatefulFuncPtr;
+template <cuda::std::size_t... Is>
+struct MakeVoidPtrStatefulFuncPtr<cuda::std::index_sequence<Is...>>
+{
+  template <cuda::std::size_t>
+  using void_ptr_t = const void*;
+  using type       = void (*)(void*, void_ptr_t<Is>..., void*); // First void* is state, last is result
+};
+
 template <typename Tag, cccl_op_t_mapping Operation, cccl_type_info_mapping RetT, cccl_type_info_mapping... ArgTs>
 struct stateless_user_operation
 {
-  __device__ decltype(RetT)::Type operator()(decltype(ArgTs)::Type... args) const
+  // Note: The user provided C f  unction (Operation.operation) must match the signature:
+  // void (void* arg1, ..., void* argN, void* result_ptr)
+  __device__ typename decltype(RetT)::Type operator()(typename decltype(ArgTs)::Type... args) const
   {
-    return reinterpret_cast<decltype(RetT)::Type (*)(decltype(ArgTs)::Type...)>(Operation.operation)(
-      std::move(args)...);
+    using Indices        = cuda::std::make_index_sequence<sizeof...(ArgTs)>;
+    using TargetCFuncPtr = typename MakeVoidPtrFuncPtr<Indices>::type;
```

Comment:
This shouldn't be necessary - something like `decltype(args, void())*...` creates the same void* pack as above, but doesn't need additional helpers (and instantiating those has a cost).

Reply:
Fixed

Reply:
Now that the use of `MakeVoidPtrFuncPtr` has been removed, its definition can be removed too.

Reply:
Also, the replacement resulted in `const` qualifier for the input parameter pointer dropped, that it changed `const void *` to `void *`.

Reply:
@oleksandr-pavlyk or @griwes  do you have recommendations for how to maintain the `const`-ness? I'm struggling a bit with syntax here.

Reply:
Never mind, I just saw Sasha's comments below